feat: multi-tenancy (#2887)
emqtt/emq-docs-cn: en_US/multi-tenancy/namespace.md

# Namespace

Starting from EMQX 5.9.0, the Namespace feature allows users to logically isolate MQTT clients, topics, authentication rules, and traffic limits within a single EMQX cluster. This feature enables scalable deployments where multiple client groups (such as business units, applications, or customers) share the same infrastructure while remaining logically separated.

::: tip Note

This feature is referred to as Namespace in EMQX 5.9, even though it follows multi-tenancy design principles.

:::

## What Is a Namespace

A Namespace is a logical boundary in EMQX used to group MQTT clients and apply isolation policies. It is identified by the special client attribute `tns` (tenant namespace), which can be extracted from connection metadata such as the username or Server Name Indication (SNI).

Namespaces can be used to control:

- Topic isolation using automatic topic prefixing (mountpoint)
- Client ID isolation to avoid clashes between tenants

> - Authentication and authorization policies per namespace

Author: I don't think we have this implemented yet

- Per-namespace client count limits and connection tracking

> - Logging metadata enrichment with namespace information

Author: I think this is not implemented as well

Author: Ignore me: there's indeed something in this vein.

You can extract the `tns` attribute from connection metadata, such as the client’s username, SNI, or other fields.

For example, to use the client's username as the namespace identifier, you can use the following configuration:

```hcl
mqtt.client_attrs_init = [{expression = username, set_as_attr = tns}]
```

### Methods for Creating a Namespace

There are two methods for creating namespaces: explicit creation and automatic creation.

1. **Explicit Namespace Creation**
   You can manually create a namespace through the Dashboard or REST API. Explicitly created namespaces can be directly managed, edited, and deleted.

   **Use Case**: This method is ideal when you need full control over which namespaces exist and wish to manage them individually.

2. **Automatic Namespace Creation via EMQX Extracting `client_attrs.tns`**
   When clients connect, EMQX can automatically create namespaces based on the client’s `client_attrs.tns` attribute. This method is usually used in scenarios where manual namespace creation is not necessary and is suitable for large-scale automated deployments.

>    **Use Case**: This approach works best for environments where clients connect dynamically, and each tenant or client requires an automatically generated, independent namespace.

Author: ```suggestion
   **Use Case**: This approach works best for environments where clients connect are trusted, and each tenant or client requires an automatically generated, independent namespace.
```

   ::: tip Note

   Automatically created namespaces cannot be edited in the Dashboard. They are generated either through configuration files or automatically based on client metadata.

   :::

> 

Author: ```suggestion

    :::: tip Note
    Automatic namespace creation is only enabled when `multi_tenancy.allow_only_managed_namespaces = false`.
    ::::
```

You can configure per-namespace rate limiters to control traffic and message flow for specific client groups. These limiters work in conjunction with EMQX's existing rate limiters for zones and listeners, depending on the type of limiter configured.

### Limiter Types

In a managed namespace, there are two types of rate limiters:

**Tenant rate limiters**: Assign tokens that are **shared** across all clients within a namespace (NS). When this type of limiter is configured, it composes with any existing zone-level rate limiters, meaning both the zone and the namespace tenant rate limiters apply to clients simultaneously.

**Client rate limiters**: Assign tokens that are **exclusive** to each client within the NS. When this type of limiter is configured, it replaces any existing listener-level rate limiters, meaning the listener rate limiters are ignored while the namespace client limiter takes effect.

Both limiter types can define limits for:

- **Message rate limits**: the maximum number of messages a client or tenant can publish over a given time period
- **Byte throughput limits**: the maximum allowed size for message payloads over time

> - **Subscription rate limits**: the maximum number of topics a client can subscribe to within a specified time window

Author: The original form was correct: messages rate limit limits subscriptions too.

```suggestion
```

Reviewer: I removed the bullet, but the "Message Publish Rate" option only say the "the maximum number of messages a tenant can **send**", it doesn't mention message "subscription".

Author: Sorry for the confusion, either that changed recently, or I confused the situations when subscriptions could be limited. 

`SUBSCRIBE` packets are actually _not_ rate limited.  :see_no_evil: :thinking:

     - **Data Publish Burst**: Allows additional bytes to be sent during bursts.
     - **Messages Publish Rate**: Limits the maximum number of messages a tenant can send per second.
     - **Messages Publish Burst**: Allows additional messages to be sent during bursts.

   - **Client Limiter**: This configuration controls traffic for individual clients. Client rate limiters are exclusive to each client, meaning the rate limit for one client won’t affect others. By default, this is disabled. If enabled, you can configure the following rate limits:

     ::: tip

     For more details on this configuration, refer to the tooltips in the Dashboard.

     :::

     - **Data Publish Rate**: Limits the bytes a client can send to EMQX per second.
     - **Data Publish Burst**: Allows additional bytes to be sent during bursts.
     - **Messages Publish Rate**: Limits the maximum number of messages a client can send per second.
     - **Messages Publish Burst**: Allows additional messages to be sent during bursts.

3. After completing the configuration, click **Create**. The new namespace will appear in the list.

> #### Manage Namespaces

Author: Nit: seems like this section is duplicated above?  i.e. there seems to be a "manage namespace" section already.

Reviewer: I updated the title to make it more specific.
